In [1]:
import pandas as pd

In [30]:
data_dir = "./data/raw_data/"
save_dir = "./data/preprocessed_data/updated_data/"

Read in the data

In [7]:
cv_table = pd.read_csv(data_dir + "updated_data/CRR.csv")
ndr_table = pd.read_csv(data_dir + "updated_data/WQR.csv")
poll_table = pd.read_csv(data_dir + "updated_data/POLL.csv")

latlong_cv = pd.read_csv(data_dir + "latlong_cv.csv")
latlong_ndr = pd.read_csv(data_dir + "latlong_ndr.csv")
latlong_pollination = pd.read_csv(data_dir + "latlong_pollination.csv")


We have three tables of data, one for each contribution of nature. We also have three tables relating each data point to a longitude and a latitude.

Let's look at the tables first!

# cv_table

In [8]:
cv_table.describe(include = 'all')

fid    country         region           UN_c           UN_1  \
count   686665.000000     685971         685971  686665.000000  686665.000000   
unique            NaN        161              8            NaN            NaN   
top               NaN  Indonesia  North America            NaN            NaN   
freq              NaN      74263         160461            NaN            NaN   
mean    343464.696723        NaN            NaN       2.483391       2.823775   
std     198006.613779        NaN            NaN       0.511871       0.573231   
min       3304.000000        NaN            NaN       1.030057       1.237033   
25%     171667.000000        NaN            NaN       2.094444       2.376177   
50%     343333.000000        NaN            NaN       2.449490       2.786279   
75%     514999.000000        NaN            NaN       2.826065       3.259844   
max     686665.000000        NaN            NaN       4.291871       4.817462   

                 UN_3           UN_5           NC_c           NC_1  \
count   686665.000000  686665.000000  686665.000000  686665.000000   
unique            NaN            NaN            NaN            NaN   
top               NaN            NaN            NaN            NaN   
freq              NaN            NaN            NaN            NaN   
mean         2.984508       3.140089       0.037279       0.036749   
std          0.617884       0.644807       0.078843       0.078211   
min          1.297790       1.311983      -1.000000      -1.000000   
25%          2.492883       2.667168       0.000000       0.000000   
50%          2.935599       3.107233       0.000000       0.000000   
75%          3.419952       3.619904       0.092263       0.066752   
max          5.000000       5.000000       0.235276       0.235276   

                 NC_3           NC_5         pop_c         pop_1  \
count   686665.000000  686665.000000  6.866650e+05  6.866650e+05   
unique            NaN            NaN           NaN           NaN   
top               NaN            NaN           NaN           NaN   
freq              NaN            NaN           NaN           NaN   
mean         0.036893       0.036704  1.126055e+03  1.373949e+03   
std          0.078365       0.078179  7.320693e+04  9.411554e+04   
min         -1.000000      -1.000000  0.000000e+00  0.000000e+00   
25%          0.000000       0.000000  4.970154e-01  4.151525e-01   
50%          0.000000       0.000000  2.177214e+01  2.147821e+01   
75%          0.071753       0.062181  1.606435e+02  1.571138e+02   
max          0.235276       0.235276  3.145269e+07  4.268553e+07   

               pop_3         pop_5  
count   6.866650e+05  6.866650e+05  
unique           NaN           NaN  
top              NaN           NaN  
freq             NaN           NaN  
mean    1.421939e+03  1.398819e+03  
std     9.292145e+04  9.330428e+04  
min     0.000000e+00  0.000000e+00  
25%     4.487666e-01  4.361424e-01  
50%     2.678465e+01  2.193834e+01  
75%     1.878266e+02  1.628649e+02  
max     4.358711e+07  4.143670e+07

This table describes the coastal risk mitigation by natural habitats.
Some explanation of the table follow:

* **fid** - the id of the geographical location that the data describes. This id can be converted to lon/lat using the latlong table.  
* **UN_cur** - the **current** (2015) unmet need of the coastal risk mitigation in this particular area. If this figure is negative then it means that we have enough, and even a surplus, in this particular area. 
* **UN_ssp1** the unmet need of the coastal risk mitigation in the year 2050, given that scenario **ssp1** is realized. 
* **ssp1, ssp3, ssp5** - the three future scenarios. It is not perfectly clear which scenraio is which here. ssp1 is most likely the sustainability scenario, purely based on the observation that the unmet need in this column is lower than for the other scenarios. The other two scenarios are fossil-fuels and regional rivalry.
* **pop_cur** - the **current** (2015) number of people in this geographic region who are affected by coastal risk mitigation. If the unmet need is high and the population is high then a lot of people are negatively affected.
* **pop_sspX** - same as the above, but in the year 2050, for the three different future scenarios.

# ndr_table

In [9]:
ndr_table.describe(include = 'all')

fid country  region          UN_c          UN_1  \
count   13215.000000   13215   13215  1.321500e+04  1.321500e+04   
unique           NaN     168       7           NaN           NaN   
top              NaN  Russia  Africa           NaN           NaN   
freq             NaN    1506    2738           NaN           NaN   
mean    24756.199924     NaN     NaN  5.439896e+06  5.696458e+06   
std     10348.534721     NaN     NaN  1.053685e+07  1.291153e+07   
min      9925.000000     NaN     NaN  0.000000e+00  0.000000e+00   
25%     15790.500000     NaN     NaN  7.382453e+05  4.839964e+05   
50%     22568.000000     NaN     NaN  2.256096e+06  1.603293e+06   
75%     32980.500000     NaN     NaN  5.625936e+06  5.526796e+06   
max     52313.000000     NaN     NaN  1.714816e+08  3.159408e+08   

                UN_3          UN_5          NC_c          NC_1          NC_3  \
count   1.321500e+04  1.321500e+04  13215.000000  13215.000000  13215.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean    6.617717e+06  5.517637e+06      0.722226      0.758806      0.712588   
std     1.410886e+07  1.121654e+07      0.155923      0.145997      0.230949   
min     0.000000e+00  0.000000e+00      0.268032      0.133294     -2.215372   
25%     5.232498e+05  4.904952e+05      0.616834      0.670287      0.612981   
50%     1.879288e+06  1.738118e+06      0.739356      0.793527      0.773301   
75%     6.390241e+06  5.948842e+06      0.845648      0.855730      0.857030   
max     2.093042e+08  2.064157e+08      1.000000      1.000000      1.000000   

                NC_5         pop_c         pop_1         pop_3         pop_5  
count   13215.000000  1.321500e+04  1.183900e+04  1.183900e+04  1.183900e+04  
unique           NaN           NaN           NaN           NaN           NaN  
top              NaN           NaN           NaN           NaN           NaN  
freq             NaN           NaN           NaN           NaN           NaN  
mean        0.741928  3.233945e+05  2.150300e+05  4.964972e+05  2.140241e+05  
std         0.217848  9.299560e+05  5.734855e+05  1.387242e+06  5.675742e+05  
min        -3.048947  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
25%         0.661652  3.529842e+03  5.168572e+03  7.882555e+03  5.465063e+03  
50%         0.794158  3.726848e+04  3.035928e+04  5.353800e+04  3.156906e+04  
75%         0.863073  2.124498e+05  1.431630e+05  3.170438e+05  1.455778e+05  
max         1.000000  1.667823e+07  8.933015e+06  2.162865e+07  8.607843e+06

This table describes the improvement of water quality by filtering of pollutants (in our case probably excessive amounts of nutrients, making the water quality bad) when water passes by natural habitats such as forests and wetlands.

The columns are the same as in the previous table, although `pop` is replaced by `rurpop`. Charlotte has confirmed that this is still the same metric though.

We also have two extra columns `country` and `region`. We will most probably not use these, especially as we don't actually have them for each table.

One thing we can immediately observe is that the count (number of entries) is much lower, the previous table had 686665 entries while this one has only 25243. This needs to be investigated. Does this mean that we don't have data for the whole earth for this particular contribution of nature?

# poll_table

In [10]:
poll_table.describe(include = 'all')

fid country   region          UN_c          UN_1  \
count    9160.000000    9160     9160  9.160000e+03  9.160000e+03   
unique           NaN     161        7           NaN           NaN   
top              NaN  Russia  Eurasia           NaN           NaN   
freq             NaN     876     1951           NaN           NaN   
mean    25270.656441     NaN      NaN  9.630618e+04  8.972863e+04   
std     10032.053267     NaN      NaN  4.348216e+05  4.136170e+05   
min      8845.000000     NaN      NaN  0.000000e+00  0.000000e+00   
25%     16672.000000     NaN      NaN  2.088596e+00  2.046068e+00   
50%     22971.500000     NaN      NaN  1.027879e+03  9.781146e+02   
75%     33446.250000     NaN      NaN  3.107308e+04  2.827017e+04   
max     51951.000000     NaN      NaN  2.322062e+07  2.249534e+07   

                UN_3          UN_5         NC_c         NC_1         NC_3  \
count   9.160000e+03  9.160000e+03  9160.000000  9160.000000  9160.000000   
unique           NaN           NaN          NaN          NaN          NaN   
top              NaN           NaN          NaN          NaN          NaN   
freq             NaN           NaN          NaN          NaN          NaN   
mean    9.375806e+04  9.990251e+04     0.342024     0.363090     0.291586   
std     4.113846e+05  4.335244e+05     0.368788     0.371759     0.354174   
min     0.000000e+00  0.000000e+00     0.000000     0.000000     0.000000   
25%     9.315915e+00  9.550000e+00     0.017078     0.019841     0.007737   
50%     1.492830e+03  1.594218e+03     0.161952     0.204912     0.089837   
75%     3.439839e+04  3.474784e+04     0.666667     0.685986     0.552810   
max     2.321814e+07  2.270223e+07     1.000000     1.000000     1.000000   

               NC_5         pop_c         pop_1         pop_3         pop_5  
count   9160.000000  9.160000e+03  6.540000e+03  6.540000e+03  6.540000e+03  
unique          NaN           NaN           NaN           NaN           NaN  
top             NaN           NaN           NaN           NaN           NaN  
freq            NaN           NaN           NaN           NaN           NaN  
mean       0.312048  4.576569e+05  9.023251e+05  1.093869e+06  8.753412e+05  
std        0.352050  1.481439e+06  2.415334e+06  2.722593e+06  2.405526e+06  
min        0.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
25%        0.014991  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
50%        0.134070  3.016314e+04  1.452440e+05  1.897702e+05  1.260224e+05  
75%        0.592607  2.764953e+05  6.747889e+05  8.782078e+05  6.261508e+05  
max        1.000000  3.046985e+07  3.778232e+07  3.752056e+07  3.774713e+07

This table describes the pollination of crops coming from natural habitats inhabiting pollinators (ex. bees).

The columns are the same as previously.

Here though we have an even lower count. What does this mean? A hypothesis could be that the resolution is simply not just not as high for the data here.

# latlong_cv

In [11]:
latlong_cv.describe(include = 'all')

fid            lat            lng
count  686665.000000  686665.000000  686665.000000
mean   343333.000000       9.346826      23.706301
std    198223.255633      32.202816     102.750279
min         1.000000     -58.480000    -180.000000
25%    171667.000000     -10.600000     -73.070000
50%    343333.000000       9.600000      12.640000
75%    514999.000000      35.450000     123.250000
max    686665.000000      62.800000     180.000000

This looks good, just a simple mapping from the **fid** to the coordinates. We have the same number of mappings as we have entrys in the `cv_table`.

# latlong_ndr

In [12]:
latlong_ndr.describe(include = 'all')

fid           lat           lng
count  64800.000000  64800.000000  64800.000000
mean   32399.500000      0.000000      0.000000
std    18706.293059     51.961123    103.923449
min        0.000000    -89.500000   -179.500000
25%    16199.750000    -44.750000    -89.750000
50%    32399.500000      0.000000      0.000000
75%    48599.250000     44.750000     89.750000
max    64799.000000     89.500000    179.500000

Here the number of entries is actually higher than the number of entries in the `ndr_table`. This is a bit strange, what does this mean?

Moreover, why do we even need to have different mappings for the different contributions of nature? Couldn't they just use the same mappings between id and coordinates? The answer to this is probably that, as suggested before, the world map resolutions of the nature's contributions are different, with coastal risk having the most detailed data.

# latlong_pollination

In [13]:
latlong_pollination.describe(include = 'all')

fid           lat           lng
count  64800.000000  64800.000000  64800.000000
mean   32399.500000      0.000000      0.000000
std    18706.293059     51.961123    103.923449
min        0.000000    -89.500000   -179.500000
25%    16199.750000    -44.750000    -89.750000
50%    32399.500000      0.000000      0.000000
75%    48599.250000     44.750000     89.750000
max    64799.000000     89.500000    179.500000

Once again the same thing, the number of entries are the same as in the `latlong_ndr` table, 64800, while the number of entries in the `poll_table` is just 9160...



# Actual data

We also take a look at the actual data, just to get a quick feel for the actual figures:

In [14]:
cv_table.head()

fid country   region      UN_c      UN_1      UN_3      UN_5  NC_c  NC_1  \
0  9715  Russia  Eurasia  2.591935  2.909349  3.112751  3.265634  -1.0  -1.0   
1  9716  Russia  Eurasia  2.865853  3.216811  3.441708  3.610748  -1.0  -1.0   
2  9717  Russia  Eurasia  2.993380  3.359955  3.594861  3.914325  -1.0  -1.0   
3  9718  Russia  Eurasia  2.720859  3.054061  3.267581  3.557960  -1.0  -1.0   
4  9719  Russia  Eurasia  2.947428  3.308376  3.539675  3.854235  -1.0  -1.0   

   NC_3  NC_5      pop_c      pop_1      pop_3      pop_5  
0  -1.0  -1.0   2.137938   2.137938   2.137938   2.137938  
1  -1.0  -1.0  11.744734  11.744734  11.744734  11.744734  
2  -1.0  -1.0  39.133438  39.133438  39.133438  39.133438  
3  -1.0  -1.0  30.200581  30.200581  30.200581  30.200581  
4  -1.0  -1.0   2.611755   2.611755   2.611755   2.611755

Nothing surprising here! The id:s start from 1 and go up. We see an unmet need for the first 5 geographical regions both currently and for the 3 different future scenarios. We also see that nobody in this region actually gets directly affected by any changes to the natural habitats mitigating coastal risk. This could be because nobody lives there, or simply perhaps because there is no coast in this area.

Let's look at the `ndr_table`.

In [22]:
ndr_table.sample(5)

fid   country         region          UN_c          UN_1  \
9207   30342  Colombia  South America  1.559986e+05  1.572014e+05   
11132  37648    Zambia         Africa  6.282378e+06  7.146704e+06   
3605   16399   Croatia        Eurasia  1.635608e+07  1.781037e+07   
5968   21146     China     North Asia  3.140675e+06  3.062480e+06   
1323   12658    Canada  North America  8.597947e+05  8.730628e+05   

               UN_3          UN_5      NC_c      NC_1      NC_3      NC_5  \
9207   1.597041e+05  1.553236e+05  0.912171  0.912045  0.910085  0.912551   
11132  1.398497e+07  8.897779e+06  0.591096  0.628992  0.089753  0.420866   
3605   2.023041e+07  1.886904e+07  0.738056  0.717846  0.676008  0.697810   
5968   3.096082e+06  3.091034e+06  0.637148  0.634187  0.642299  0.642883   
1323   8.707451e+05  8.628394e+05  0.842474  0.842526  0.840467  0.841916   

               pop_c          pop_1          pop_3          pop_5  
9207    59373.214844   69951.921875  102215.031250   69663.898438  
11132  272483.531250   91951.632812  823606.687500   89484.148438  
3605   801668.687500  265764.656250  742937.875000  266483.437500  
5968     7866.768066    3685.388916    7556.784668    3685.388916  
1323     3277.998535    3955.514648    2459.737061    5301.553711

We note that the **fid** goes above the maximum number of entries in the table (which was 25243). 

Same for `poll_table`:

In [20]:
poll_table.sample(5)

fid                           country         region         UN_c  \
8504  41891                            Brazil  South America     0.120175   
4023  20969                     United States  North America    15.745658   
3079  18552                             Italy        Eurasia    99.313338   
7286  35123  Democratic Republic of the Congo         Africa    12.414923   
1062  13916                            Russia        Eurasia  5277.380393   

             UN_1         UN_3         UN_5      NC_c      NC_1      NC_3  \
8504     0.070839     0.227290    23.659318  0.940778  0.959267  0.848204   
4023     8.322329    22.098337    19.033267  0.307840  0.625999  0.113403   
3079   106.374762   107.054507   119.974196  0.677887  0.654903  0.653196   
7286    13.005741    44.692844    41.985839  0.945499  0.940998  0.940399   
1062  6264.574134  5013.450385  5266.725183  0.038437  0.058634  0.078944   

          NC_5         pop_c         pop_1         pop_3         pop_5  
8504  0.743174  1.010766e+06  1.128160e+06  1.325037e+06  1.118941e+06  
4023  0.221615  1.873246e+05  2.182784e+05  1.632682e+05  2.668269e+05  
3079  0.612674  4.730262e+05  5.285684e+05  4.147921e+05  5.852842e+05  
7286  0.943868  4.750586e+05  9.677742e+05  1.264010e+06  9.652922e+05  
1062  0.040211  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00

And for an example lat_long-table let's check out `latlong_cv`:

In [23]:
latlong_cv.sample(5)

fid    lat     lng
237964  237965  -3.54  102.08
296358  296359   1.81  -50.28
115869  115870 -23.78  151.28
346599  346600  10.07  106.68
221805  221806  -5.04  137.56

# Preprocessing

The next step is to perform the preprocessing (if needed) and save the data in new files. 
This might need to be done in several iteration as we don't know at the moment exactly how we will use the data. One thing that will be good to have done already is to join the lat/lon tables with the data tables.

## Join lat/lon tables with the data tables

Start with `cv_table`:

In [24]:
cv_table_out = pd.merge(cv_table, latlong_cv, on='fid')
cv_table_out.sample()

fid    country         region      UN_c      UN_1      UN_3  \
361550  361551  Venezuela  South America  1.873185  2.004145  2.102579   

            UN_5      NC_c      NC_1      NC_3      NC_5      pop_c  \
361550  2.182247  0.156567  0.156567  0.156567  0.156567  47.169404   

            pop_1      pop_3      pop_5    lat    lng  
361550  61.905341  77.019355  62.328426  10.83 -68.33

`fid`, `country` or `region` column is not needed, drop this.

In [25]:
cv_table_out.drop(columns=['fid', 'country', 'region'], inplace=True)
cv_table_out.sample()

UN_c      UN_1      UN_3      UN_5  NC_c  NC_1  NC_3  NC_5  \
275644  2.492883  2.993795  3.140836  3.259844   0.0   0.0   0.0   0.0   

            pop_c      pop_1      pop_3      pop_5   lat    lng  
275644  26.405259  35.668626  63.645284  29.928352  0.96  43.84

Looks good! Let's do the same for `ndr_table` and `poll_table`:

In [26]:
ndr_table_out = pd.merge(ndr_table, latlong_ndr, on ='fid')

# Remove the country and region columns, which we will probably not use
ndr_table_out.drop(columns=['fid', 'country', 'region'], inplace=True) 

ndr_table_out.sample()

UN_c        UN_1         UN_3        UN_5      NC_c      NC_1  \
10423  2130958.25  997782.125  2023996.625  1330933.25  0.758829  0.878003   

           NC_3      NC_5        pop_c         pop_1      pop_3         pop_5  \
10423  0.770934  0.849371  1064130.125  66259.414062  2829447.5  66148.890625   

       lat   lng  
10423 -6.5  21.5

In [27]:

poll_table_out = pd.merge(poll_table, latlong_pollination, on='fid')
poll_table_out.drop(columns=['fid', 'country', 'region'], inplace=True)

poll_table_out.sample()

UN_c         UN_1        UN_3        UN_5      NC_c  \
4738  3632.010023  2323.369815  1466.55897  5470.18321  0.000344   

              NC_1          NC_3      NC_5        pop_c     pop_1  \
4738  1.014512e-16  7.882924e-17  0.001292  1070908.375  783985.0   

            pop_3      pop_5   lat    lng  
4738  979571.1875  785328.25  24.5  100.5

## Dropping rows without any inhabitants
We are interested in the number of affected people. Therefore, we can drop the data rows where there the population is 0 (currently, and for all of the 3 scenarios in 2050), e.g. places which are inhabited.

In [28]:
cols = ['pop_c', 'pop_1', 'pop_3', 'pop_5']
print("Number of rows before: " + str(cv_table_out.shape[0]))
cv_table_out = cv_table_out.loc[(cv_table_out[cols]!=0).any(axis=1)]
print("Number of rows after: " + str(cv_table_out.shape[0]))

Number of rows before: 686665
Number of rows after: 635204


Doing the same for the ndr_table and the poll table:

In [29]:
print("Number of rows in ndr_table before: " + str(ndr_table_out.shape[0]))
ndr_table_out = ndr_table_out.loc[(ndr_table_out[cols]!=0).any(axis=1)]
print("Number of rows in ndr_table after: " + str(ndr_table_out.shape[0]))

print("")

print("Number of rows in poll_table before: " + str(poll_table_out.shape[0]))
poll_table_out = poll_table_out.loc[(poll_table_out[cols]!=0).any(axis=1)]
print("Number of rows in poll_table after: " + str(poll_table_out.shape[0]))

Number of rows in ndr_table before: 13215
Number of rows in ndr_table after: 13215

Number of rows in poll_table before: 9160
Number of rows in poll_table after: 7737


A lot of data is lost doing this, especially proportionately in the case of the ndr and the poll datasets.

This is fine though since, as stated before, we are only interested in the number of affected people.

*In this iteration 2, we note that the ndr_table was already cleaned of empty population rows. This strenghtens the choice of removing them, and we continue doing so in this updated data preprocessing*.

## Subsampling data

The cv dataset is a bit too large to be used as it is (will have too long loading times).
Therefore we subsample it to be smaller:

In [31]:
cv_table_out = cv_table_out.sample(30000)

Now we are ready to save the files!

# Saving data

Doing a final reset of the indexes:

In [32]:
cv_table_out.reset_index(drop=True, inplace=True)
ndr_table_out.reset_index(drop=True, inplace=True)
poll_table_out.reset_index(drop=True, inplace=True)

In [33]:
cv_table_out.to_csv(save_dir + "cv_table_preprocessed.csv", index=False)
ndr_table_out.to_csv(save_dir + "ndr_table_preprocessed.csv", index=False)
poll_table_out.to_csv(save_dir + "poll_table_preprocessed.csv", index=False)